qiskit/circuit/add_control.py

```diff
@@ -218,7 +217,8 @@ def control(
     if isinstance(operation, controlledgate.ControlledGate):
         operation.ctrl_state = original_ctrl_state
         new_num_ctrl_qubits = num_ctrl_qubits + operation.num_ctrl_qubits
-        new_ctrl_state = operation.ctrl_state << num_ctrl_qubits | ctrl_state
+        # TODO: what if ctrl_state is str?
```

Comment:
`ctrl_state` can be string but is not converted to int anywhere

Reply:
It looks like both `str` and `None` are not supported here. The `ControlledGate` itself already does the string-or-None to int conversion so it's likely not needed here and we can remove it from the typehint.

Reply:
Neither `add_control` function nor control method take care of string case:
```
>>> from qiskit.circuit.library import CRYGate
>>> CRYGate(1/2).control(5, ctrl_state=1)
Instruction(name='c6ry_o33', num_qubits=7, num_clbits=0, params=[0.5])
>>> CRYGate(1/2).control(5, ctrl_state='00001')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.11/site-packages/qiskit/circuit/gate.py", line 121, in control
    return add_control(self, num_ctrl_qubits, label, ctrl_state)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/qiskit/circuit/add_control.py", line 59, in add_control
    cgate = control(operation, num_ctrl_qubits=num_ctrl_qubits, label=label, ctrl_state=ctrl_state)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/qiskit/circuit/add_control.py", line 221, in control
    new_ctrl_state = operation.ctrl_state << num_ctrl_qubits | ctrl_state
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
TypeError: unsupported operand type(s) for |: 'int' and 'str'
>>> 
```

Reply:
Then I'd say we can safely remove it from both functions, as they are not getting `str` or `None` as input anyways 😄

Reply:
I think what happens is that for many standard gates, this code path is not executed, e.g.,
```
>>> from qiskit.circuit.library import CCXGate
>>> CCXGate().control(1, ctrl_state=1)
Instruction(name='mcx', num_qubits=4, num_clbits=0, params=[])
>>> CCXGate().control(1, ctrl_state='1')
Instruction(name='mcx', num_qubits=4, num_clbits=0, params=[])
```
I wouldn't safely assert that it is not used. Isn't the correct fix to add `_ctrl_state_to_int` somewhere?

Reply:
@1ucian0 @ewinston what do you think?

Reply:
> Isn't the correct fix to add _ctrl_state_to_int somewhere?

It is already called inside the `ControlledGate`: https://github.com/Qiskit/qiskit-terra/blob/789d588f14d873ad4b3c22ac3e6cbada856239da/qiskit/circuit/controlledgate.py#L206 🙂

Reply:
Oh yeah, sorry I didn't realize the problem was in the `Gate.control` function 😄 That's indeed a bug and we should add the conversion to `add_control` and `control`.  We could also add the conversion inside the `Gate.control` method, but if we add them to `add_control/control` we can leave the type hints as they were intended so that seems like the better solution.

We should definitely add a test to cover these cases as there seem to be none in place, too!

Thanks for the code example 👍🏻